---
title:  Unsupervised Clustering of Breast Cancer Samples              
author: Lela Makharadze 
date: 29.11.2020
output: html_document
---

Please note:  the code considers that all data sets are obtained afresh once respective chunks are run. 
Thus, no files are attached to the project (working directory can be freely set).

```{r setup, loading packages}
library(TCGAbiolinks)
library(SummarizedExperiment)
library(ggplot2)
library(knitr)
library(dplyr)
library(tidyverse)
library(EDASeq)
library(knitr)
library(limma)
library(GO.db)
library(org.Hs.eg.db)
library(ReactomePA)
library(ELMER)
library(MultiAssayExperiment)
library(mygene)
library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
library(GGally)
library(ggpubr)
library(patchwork)
library(knitr)
library(rmarkdown)
library(kableExtra)
library("EnsDb.Hsapiens.v79")
library(ReactomePA)
library(reactome.db)
library(magrittr)
library(tinytex)
library(preprocessCore)
```



#### Transcriptome Data Access

Using *TCGAbiolinks* package, RNAseq data of 500 samples were obtained from "TCGA-BRCA" project followed by the essential data normalization and transformation steps. 
```{r Transcriptome data access and filtering}

#--------------------------------------------------------------------------------------
                                          #gene expression data
#--------------------------------------------------------------------------------------
samples500 <- matchedMetExp("TCGA-BRCA", n = 500)
#matchedMetExp will consider the same data for later during methylation data access

query <- GDCquery(project = "TCGA-BRCA",
                  data.category = "Gene expression",
                  data.type = "Gene expression quantification",
                  experimental.strategy = "RNA-Seq",
                  platform = "Illumina HiSeq",
                  file.type = "results",
                  legacy = TRUE,
                  barcode = samples500)
GDCdownload(query)
geneexpr <- GDCprepare(query)
ExpMatrix <- assay(geneexpr) #gene expression matrix

#NORMALIZATION AND FILTERING
dataPrep <- TCGAanalyze_Preprocessing(object=geneexpr,
                                          cor.cut = 0.6,
                                          datatype = "raw_count",
                                          filename = "AAIC.png") 
#19947 genes 500 samples - neither samples nor genes filtered


dataNorm <- TCGAanalyze_Normalization(tabDF = dataPrep,
                                      geneInfo = TCGAbiolinks::geneInfo,
                                      method="gcContent") #19866 genes  ~100 genes were removed

dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm,
                                  method = "quantile",
                                  qnt.cut =  0.25) #14899 genes ~500 genes were removed


``` 


#### Principle Component Analysis and *k*-means Clustering


```{r filtering out normal samples}
#Organizing data for PCA
ExpMatrixlog2<-as.data.frame(log(dataFilt+1)) #all 500 samples
ExpMatrixLog2T<- as.data.frame(t(ExpMatrixlog2))

#filtering normal samples out of tumor population
samplesNT <- TCGAquery_SampleTypes(barcode = colnames(dataNorm), 
                                   typesample = "NT") #normal tissue

samplesTP <- TCGAquery_SampleTypes(barcode = colnames(dataNorm), 
                                   typesample =c("TP")) #tumor patient

onlytumor <- as.data.frame(ExpMatrixLog2T[samplesTP,])#transcriptome dataframe with 449 tumor samples
```

```{r PCA and k-means}
#--------------------------------------------------------------------------------------
                          #PCA and k-means clustering 
#--------------------------------------------------------------------------------------

#my function  for  PCA in 2 steps:
#(can be written in one step but for the specific plotting cases 2 steps were needed)
PCA <- function(data) {
pc <- prcomp(data)
pc.sum <- summary(pc)$importance
varExp <- round(pc.sum[2, ] * 100, 2)
pcs <- as.data.frame(pc$x)
fourdata <-list(pc, pc.sum, varExp, pcs)
return(fourdata)
}

PCAplot <- function(x){ 
  ggplot(data = x[[4]], aes(x = PC1, y = PC2 ))+geom_point()+ xlab(paste("PC1 (", x[[3]][1], "%)"))  + ylab(paste("PC2 (", x[[3]][2], "%)"))+theme_bw()+theme(legend.position = "none")
}



# ATTENTION EACH TIME U RUN KMEANS CLUSTER NUMBERING IS DIFERENT, thus colors should be re-matched manually during plotting . 
# PCA and k-means for all 500 samples
PCAall <-PCA(ExpMatrixLog2T)  
resultsAll <- kmeans(PCAall[[4]][, 1:5], 3)
resultsAll$cluster <-as.factor(resultsAll$cluster) #essential step for later plot labeling. 

kplotAll <- ggplot(data = PCAall[[4]], aes(x = PC1, y = PC2, color=resultsAll$cluster))+ xlab(paste("PC1 (", PCAall[[3]][1], "%)"))  + ylab(paste("PC2 (", PCAall[[3]][2], "%)"))+geom_point()+theme_bw()+theme(legend.position = "none") +scale_color_manual(values = c( "red","purple", "deepskyblue1" ))

#k-means for all 500 samples overlayed with normal samples
kplotAllHighlight <- kplotAll+ geom_point(data=PCAall[[4]][samplesNT,], aes(x = PC1, y = PC2), color="blue")+theme(legend.position = "none")

#PCA and k-means for only tumor samples
PCATP <- PCA(onlytumor)
PCATPplot<-PCAplot(PCATP)
resultsTP <- kmeans(PCATP[[4]][, 1:6], 2)  
resultsTP$cluster<- as.factor(resultsTP$cluster)
kplotTP <-ggplot(data = PCATP[[4]], aes(x = PC1, y = PC2, color=resultsTP$cluster ))+ xlab(paste("PC1 (", PCATP[[3]][1], "%)"))  + ylab(paste("PC2 (", PCATP[[3]][2], "%)"))+geom_point()+theme_bw()+theme(legend.position = "none") +scale_color_manual(values = c("purple","red"))

#all four plots merged
fourplots <-cowplot::plot_grid(kplotAll, kplotAllHighlight, PCATPplot, kplotTP, align = "hv", nrow = 2,  labels = c('A', 'B',  'C', 'D')) 

print(fourplots)

```



#### Differential Gene Expression Analysis (DEA)

Using *limma* package DEA was conducted. The analysis allowed identification of top up- and down-regulated genes depicted with the volcano plot.
``` {r DEA}

#organizing data for DEA
onlytumor$cluster <- as.numeric(resultsTP$cluster) #assigning clusters to  samples
onlytumorUN <- as.data.frame(t(onlytumor)) #transposed back -this way required for limma

which(rownames(onlytumorUN)=="cluster" ) # 14900 = cluster row

#--------------------------------------------------------------------------------------
                                      #LIMMA
#--------------------------------------------------------------------------------------
groups <- as.factor(onlytumorUN[14900,]) # I provide 2 groups=clusters
designs <- model.matrix(~0 + groups, data=onlytumorUN[-14900,]) #and exclude cluster IDs from calculation

#ATTENSION: depending  on cluster numbering you choose which cluster you want to use a reference thus subtract from there

#Contrasts matrix
ContrastMat <- makeContrasts(difference = groups1-groups2, levels = designs)

fits <- lmFit(onlytumorUN[-14900,], designs)

vfits <- contrasts.fit(fits, contrasts=ContrastMat)

efits <- eBayes(vfits)

volcanoplot(efits, highlight = 20,  names = rownames(efits$coefficients), hl.col = "blue")

```


#### Pathway Enrichement Analysis (PEA)


```{r reactome}
#change symbol names to ENTREZ ID
top.table <- topTable(efits, number = 20) #top 20 genes from limma

toptableIDs <-clusterProfiler::bitr(rownames(top.table), 'SYMBOL', 'ENTREZID', OrgDb=org.Hs.eg.db) #ENTREZIDs should be provided

paths <- enrichPathway(gene=toptableIDs[,2], pvalueCutoff=0.05, readable=T)
pathsDF<- as.data.frame(paths)

#table output
patways <-subset(pathsDF[1:5, c("Description", "GeneRatio", "geneID")])
rownames(output) <-NULL
kbl(pathways,  caption = "Table 1. Reactome assigned top DEGs to major ESR1 and ERBB4 mediated pathways") %>%
kable_paper(full_width = F) %>%kable_styling()



```



#### Methylation Data Access

Using *MultiAssayExperiment* package, methylation data with the matched samples from the gene expression data was obtained and aligned with the transcriptome data.

```{r methylome data}

#--------------------------------------------------------------------------------------
                                       #methylation data
#--------------------------------------------------------------------------------------


querymeth500 <- GDCquery(project = "TCGA-BRCA",
                 data.category = "DNA methylation",
                 platform = "Illumina Human Methylation 450",
                 legacy = TRUE,
                 barcode = samples500) #sample500 is the variable from transcriptome code

GDCdownload(querymeth500)
met500 <- GDCprepare(querymeth500, save = FALSE)
met500 <- met500[rowSums(is.na(assay(met500))) == 0,] # remove probes with NA
metsub <- subset(met500, subset = as.character(GenomicRanges::seqnames(met500)) %in% c( "chr2", "chr4", "chr6", "chr7", "chr21")) #chromosomes chosen based on some top key DEGs' chromosomal locations

#--------------------------------------------------------------------------------------
                            #MATCHING TRANSCRIPTOME AND METHYLOME DATA SETS
#--------------------------------------------------------------------------------------

rownames(exp) <- values(exp)$ensembl_gene_id  #"createMAE" should receive ENSEMBLE IDs (ENSG)
mae <- createMAE(exp = assay(geneexpr)[,samplesTP],  #so only tumor samples this time :)
                 met = metsub,
                 save = TRUE,
                 linearize.exp = TRUE,
                 save.filename = "mae.rda",
                 met.platform = "450K",
                 genome = "hg19",
                 TCGA = TRUE)

maematrlist <- MultiAssayExperiment::assays(mae) 
widematrix <-  wideFormat(mae)
maematrices <- as.matrix(widematrix) #gene expression and methylation matrices together
genesprobes <- as.data.frame(maematrices)
genesprobes1<- genesprobes[,-1] # sample IDs were in the first column not as row names
rownames(genesprobes1) <- genesprobes[,1] # made them as rownames
names(genesprobes1) = gsub(pattern = "DNA.methylation_", replacement = "", x = names(genesprobes1)) #removed extra string in probe name
names(genesprobes1) = gsub(pattern = "Gene.expression_", replacement = "", x = names(genesprobes1)) #and gene name


```


#### Methylome Analysis of Breast Cancer Subtypes 

Unsupervised ML was also applied to the methylome data.
```{r unspurevised clustering of methylome}
#organizing methylome data for PCA
#data has already been transposed 
genesprobes2 <- as.data.frame(sapply(genesprobes1, as.numeric)) #this conversion is essential for PCA. 
probesonly <- genesprobes2[, -grep("ENS", colnames(genesprobes2))]  # df only with probes, genes omitted. ~92K probes

#PCA of methylome
PCAprobes <- PCA(probesonly)
PCAprobesplot <- PCAplot(PCAprobes)
print(PCAprobesplot)

```




Methylation levels were tested for differentially expressed genes to find a correlation between gene expression status and methylation of CpGs associated with them. To this end, several additional operations were required to find matching probes for the genes of interest.

```{r matching probes with genes}
#probe and gene annotation data
 annots = getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
 annots <- as.matrix(annots)

#genes of interest to find matching probes for
tops <- topTable(efits, number=Inf) # get all genes from DEA
DOWNS <- tops[which(tops$logFC > 0), ] [1:30,] # up-reg top 30
UPS <- tops[which(tops$logFC < 0), ] [1:30,] # down-reg top 30
UN<- tops[which(tops$P.Value > 0.09 & tops$logFC < 0.5 & tops$logFC > 0) & tops$t < 1 & tops$t > 0,] %>% drop_na()  #this is how I define non-differentially expressed genes

#first I filter data by chromosomes and then I pick gene names (string, because in annots several genes are pasted in one  column) from  upregulated gene set.  So I get a filtered annotation matrix.
ENSvecUPfinal <- dplyr::filter(annots, annots$chr==c( "chr2", "chr4", "chr6", "chr7", "chr21")) %>% dplyr::filter(grepl(paste(rownames(UPS), collapse="|"), UCSC_RefGene_Name))

#I turn it into a vector that contains only row names(=probe names) of that filtered annots matrix
vecUPS <- rownames(ENSvecUPfinal)


#The same should be performed for down-regulated and unchanged gene sets:

ENSvecDOWNfinal <- dplyr::filter(annots, annots$chr==c( "chr2", "chr4", "chr6", "chr7", "chr21")) %>% dplyr::filter(grepl(paste(rownames(DOWNS), collapse="|"), UCSC_RefGene_Name))
vecDOWNS <- rownames(ENSvecDOWNfinal)

ENSvecUNfinal <- dplyr::filter(annots, annots$chr==c( "chr2", "chr4", "chr6", "chr7", "chr21"))  %>% dplyr::filter(grepl(paste(rownames(UN[800:860, ]), collapse="|"), UCSC_RefGene_Name))
vecUN <- rownames(ENSvecUNfinal)

#Now I subset my original probes dataframe based on the variables from an external annots data.

#non-differential gene probes
MatchitUN <-intersect(colnames(probesonly), vecUN) #making sure all probes from annots can be found in my df
MatchselectUN <- select_(probesonly, .dots = MatchitUN) 

#Upregulated gene probes
MatchitUP <-intersect(colnames(probesonly), vecUPS)
MatchselectUP <- select_(probesonly, .dots = MatchitUP) 

#Downregulated gene probes
MatchitDOWN <- intersect(colnames(probesonly), vecDOWNS)
MatchselectDOWN  <- select_(probesonly, .dots = MatchitDOWN) 

#binding all probes and  calcuating means per sample for each up-, down-regulated and unchanged gene probes.

MyProbes <- cbind(MatchselectUN, MatchselectUP, MatchselectDOWN)
MyProbes <- as.data.frame(sapply(MyProbes, as.numeric))
MyProbes$MeanProbeUN <- rowMeans(MyProbes[ , colnames(MatchselectUN)])
MyProbes$MeanProbeDOWN <- rowMeans(MyProbes[, colnames(MatchselectDOWN)])
MyProbes$MeanProbeUP <- rowMeans(MyProbes[, colnames(MatchselectUP)])


#finally correct assignment of samples to clusters was possible
MyProbes$cluster <- as.factor(onlytumor$cluster)

```




Methylation level of up- and down-regulated genes demonstrated significant variation between ER- and ER+ clusters, whereas probes corresponding to the non-differentially expressed genes showed no variance.


```{r  boxplots}
#assigning clusters to ER- and ER+ groups
#cluster numbering  - check if the assignment is right as kmeans changes cluster numbers
MyProbes$cluster <- as.character(MyProbes$cluster)
MyProbes$cluster[MyProbes$cluster == "2"] <- "ER+"
MyProbes$cluster[MyProbes$cluster == "1"] <- "ER-"


UNbox <- MyProbes[, -161] %>%ggplot(aes(x = cluster, y = MeanProbeUN, fill=cluster)) #had a probe duplication at 161 and omitted
a <-UNbox + geom_boxplot()+geom_jitter(color="black", size=0.4, alpha=0.9) + geom_signif(test ="t.test", comparisons = list(c("ER-", "ER+")), map_signif_level=FALSE)+scale_fill_manual(values=c( "purple","red" ))+ xlab(NULL) + ylab("Mean DNA Methylation (62 probes)")+theme_bw()+scale_y_continuous(limits = c(0.2, 0.7), breaks = c(0.2, 0.4, 0.6))


UPbox <- MyProbes[, -161] %>%ggplot(aes(x = cluster, y = MeanProbeUP, fill=cluster)) 
b<-UPbox + geom_boxplot()+geom_jitter(color="black", size=0.4, alpha=0.9)+ geom_signif(test ="t.test", comparisons = list(c("ER+", "ER-")), map_signif_level=TRUE)+scale_fill_manual(values=c("purple","red" ))+ theme(legend.position="none") +xlab(NULL) + ylab("Mean DNA Methylation (59 probes)")+theme_bw()+theme(legend.position = "none")+scale_y_continuous(limits = c(0.2, 0.7), breaks = c(0.2, 0.4, 0.6))


Downbox <- MyProbes[, -161] %>%ggplot(aes(x = cluster, y = MeanProbeDOWN, fill=cluster))
c<- Downbox + geom_boxplot()+geom_jitter(color="black", size=0.4, alpha=0.9) + geom_signif(test ="t.test", comparisons = list(c("ER+", "ER-")), map_signif_level=TRUE)+scale_fill_manual(values=c("purple","red" )) + theme(legend.position="none") +xlab(NULL) + ylab("Mean DNA Methylation (66 probes)")+theme_bw()+theme(legend.position = "none")+scale_y_continuous(limits = c(0.2, 0.7), breaks = c(0.2, 0.4, 0.6))

methboxp <-cowplot::plot_grid(b, c, a,  nrow = 1, labels = c('A', 'B', "C"), rel_widths=c(1, 1, 1.2))
print(methboxp)

```








